File: android/app/src/main/kotlin/de/godisch/potillus/domain/ChartBucketing.kt
/* vim: set et ts=4:
 * =============================================================================
 * Libellus Potionis "Potillus" -- Privacy-Friendly Alcohol Tracker
 * Copyright (c) 2026 Martin A. Godisch <android@godisch.de>
 * =============================================================================
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * =============================================================================
 */
package de.godisch.potillus.domain

import de.godisch.potillus.domain.model.DaySummary
import java.time.LocalDate

// =============================================================================
// ChartBucketing.kt – Continuous time-axis series for the consumption chart
// =============================================================================
//
// WHY THIS EXISTS:
//   The daily-summary queries return only days that HAVE entries. A bar chart
//   built directly from them has no notion of abstinent days: gaps simply
//   vanish and the x-axis is a list of drink days rather than a real time axis.
//
//   To draw a proper time axis (and to show abstinent days), the sparse
//   summaries must be expanded into a GAP-FREE series that covers every day in
//   the period. For long periods (a whole year) one bar per day is unreadable,
//   so adjacent days are aggregated into fixed-width buckets (weeks or months).
//
// SEMANTICS OF A BUCKET (see [ChartBucket]):
//   - A bucket's value is the MEAN grams of pure alcohol PER CALENDAR DAY inside
//     the bucket (sum of the bucket's days ÷ number of those days). For a DAILY
//     bucket that is simply the day's own total, so the daily chart is unchanged
//     in meaning. Using a per-day average keeps the dashed DAILY-LIMIT reference
//     line directly comparable across all granularities (a weekly bar above the
//     line means the week averaged more than the daily limit).
//   - A bucket is "abstinent" when the entire bucket contained 0 g. Abstinent
//     buckets carry no bar; the renderer marks them with a small green tick so
//     "recorded, zero consumption" is visually distinct from a tiny bar.
//
// This object is deliberately Android-free (pure java.time + plain data) so it
// is unit-testable on the JVM and shared by BOTH the on-screen chart
// (StatsViewModel) and the PDF export (PdfReportData).
// =============================================================================

/** Time granularity of one bar in the consumption-over-time chart. */
enum class ChartGranularity { DAILY, WEEKLY, MONTHLY }

/**
 * One column of the consumption-over-time chart.
 *
 * @param labelDate   The bucket's first calendar day as "YYYY-MM-DD". The
 *                    calling screen turns this into a short axis label
 *                    (weekday, day-of-month, month name, …), so this type stays
 *                    locale-agnostic.
 * @param avgPerDay   Mean grams of pure alcohol per calendar day in the bucket
 *                    (see file header). Equals the day's total for DAILY buckets.
 * @param isAbstinent True when the whole bucket had zero consumption; rendered
 *                    as a green tick instead of a bar.
 */
data class ChartBucket(
    val labelDate: String,
    val avgPerDay: Double,
    val isAbstinent: Boolean
)

object ChartBucketing {

    /**
     * Picks a sensible granularity for an ARBITRARY span (used by the PDF export,
     * whose date range is whatever the user chose, not WEEK/MONTH/YEAR):
     *   - ≤ 35 days  → one bar per day,
     *   - ≤ 366 days → one bar per week (at most ~53 bars),
     *   - otherwise  → one bar per month.
     *
     * @param days Number of calendar days in the (inclusive) span.
     */
    fun granularityForSpan(days: Int): ChartGranularity = when {
        days <= 35  -> ChartGranularity.DAILY
        days <= 366 -> ChartGranularity.WEEKLY
        else        -> ChartGranularity.MONTHLY
    }

    /**
     * Expands sparse [summaries] (days WITH entries only) into a continuous,
     * gap-free list of buckets covering the inclusive range [from]..[to].
     *
     * Missing days contribute 0 g, so abstinent days appear as zero-value
     * buckets. Buckets are clamped to the period end, so the last week/month may
     * be shorter than a full week/month and its average is computed over the
     * days that actually fall inside the period.
     *
     * @param summaries   Daily totals for days that have entries ("YYYY-MM-DD").
     * @param from        Inclusive period start ("YYYY-MM-DD").
     * @param to          Inclusive period end   ("YYYY-MM-DD").
     * @param granularity Bucket width (see [granularityForSpan]).
     * @return            Buckets in chronological order, or empty when from > to.
     */
    fun bucketize(
        summaries: List<DaySummary>,
        from: String,
        to: String,
        granularity: ChartGranularity
    ): List<ChartBucket> {
        val start = LocalDate.parse(from, DayResolver.DATE_FORMATTER)
        val end   = LocalDate.parse(to, DayResolver.DATE_FORMATTER)
        if (start.isAfter(end)) return emptyList()

        // O(1) lookup of a day's total; days not present here are abstinent (0 g).
        val gramsByDate: Map<String, Double> = summaries.associate { it.date to it.totalGrams }
        // One day past the period end; used as an exclusive upper bound when summing.
        val endExclusive = end.plusDays(1)

        val buckets = ArrayList<ChartBucket>()
        var bucketStart = start
        while (!bucketStart.isAfter(end)) {
            // Natural (un-clamped) end of this bucket, exclusive.
            val naturalEndExclusive: LocalDate = when (granularity) {
                ChartGranularity.DAILY   -> bucketStart.plusDays(1)
                ChartGranularity.WEEKLY  -> bucketStart.plusWeeks(1)
                // Month buckets snap to the 1st of the next month so successive
                // buckets align to calendar months even if the first one starts
                // mid-month.
                ChartGranularity.MONTHLY -> bucketStart.withDayOfMonth(1).plusMonths(1)
            }
            // Never let a bucket run past the period.
            val cappedEndExclusive = if (naturalEndExclusive.isAfter(endExclusive))
                endExclusive else naturalEndExclusive

            var sum = 0.0
            var dayCount = 0
            var day = bucketStart
            while (day.isBefore(cappedEndExclusive)) {
                sum += gramsByDate[DayResolver.formatDate(day)] ?: 0.0
                dayCount++
                day = day.plusDays(1)
            }

            buckets.add(
                ChartBucket(
                    labelDate   = DayResolver.formatDate(bucketStart),
                    avgPerDay   = if (dayCount > 0) sum / dayCount else 0.0,
                    isAbstinent = sum == 0.0
                )
            )

            bucketStart = cappedEndExclusive
        }
        return buckets
    }
}

File: ios/PotillusKit/Sources/PotillusKit/Domain/ChartBucketing.swift
// vim: set et ts=4:
// =============================================================================
// Libellus Potionis - Privacy-Friendly Alcohol Tracker
// Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
// =============================================================================
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program.  If not, see <https://www.gnu.org/licenses/>.
//
// In addition, as permitted by section 7 of the GNU General Public License,
// this program may carry additional permissions; any such permissions that
// apply to it are stated in the accompanying COPYING.md file.
//
// =============================================================================

import Foundation

// =============================================================================
// ChartBucketing.swift – turning sparse day totals into chart bars
// =============================================================================
//
// A faithful Swift port of the Android `domain/ChartBucketing.kt`.
//
// The database stores only days that HAVE entries. A chart, however, needs a
// continuous axis: an abstinent Tuesday must appear as a zero-height bar, not as
// a missing one. Bucketing expands the sparse data into a gap-free series, and —
// for longer spans — aggregates days into weeks or months so the bar count stays
// readable.
//
// All calendar arithmetic runs on a UTC-pinned calendar, for the same reason as
// in `DayResolver`: bucket boundaries must be pure calendar-day computations, so
// a device time zone or a DST transition can never shift a bar.
// =============================================================================

/// Bucket width for a chart.
public enum ChartGranularity: String, Sendable, Equatable, Codable {
    case daily = "DAILY"
    case weekly = "WEEKLY"
    case monthly = "MONTHLY"
}

/// One bar of a chart.
public struct ChartBucket: Sendable, Equatable {

    /// The bucket's first calendar day, `yyyy-MM-dd`; used as the axis label.
    public let labelDate: String

    /// Mean grams of alcohol per day over the days this bucket actually covers.
    public let avgPerDay: Double

    /// True when the bucket was recorded fully alcohol-free AND is a completed
    /// period. A bucket still containing the in-progress day never qualifies.
    public let isAbstinent: Bool

    public init(labelDate: String, avgPerDay: Double, isAbstinent: Bool) {
        self.labelDate = labelDate
        self.avgPerDay = avgPerDay
        self.isAbstinent = isAbstinent
    }
}

public enum ChartBucketing {

    /// A Gregorian calendar pinned to UTC, so bucket boundaries are
    /// zone-independent. See `DayResolver` for why days are anchored at noon.
    private static let utcCalendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }()

    /// Picks a sensible granularity for an ARBITRARY span (used by the PDF
    /// export, whose date range is whatever the user chose, not week/month/year):
    /// up to 35 days one bar per day, up to 366 days one bar per week (at most
    /// ~53 bars), otherwise one bar per month.
    ///
    /// - Parameter days: Number of calendar days in the inclusive span.
    public static func granularityForSpan(days: Int) -> ChartGranularity {
        if days <= 35 { return .daily }
        if days <= 366 { return .weekly }
        return .monthly
    }

    /// Expands sparse `summaries` (days WITH entries only) into a continuous,
    /// gap-free list of buckets covering the inclusive range `from...to`.
    ///
    /// Missing days contribute 0 g, so abstinent days appear as zero-value
    /// buckets. Buckets are clamped to the period end, so the last week or month
    /// may be shorter than a full one, and its average is taken over the days that
    /// actually fall inside the period.
    ///
    /// **The in-progress day ("today in superposition").** When `inProgressDay` is
    /// given — the current logical day, which on screen equals `to` — the bucket
    /// containing it applies the app's per-day rule in two distinct ways:
    ///
    /// 1. **Average.** While the day is not yet a drink day it is dropped from the
    ///    divisor, so the average is taken over completed days only. The bucket's
    ///    sum already excludes the empty day, so only the divisor changes. This
    ///    mirrors `DayResolver.effectivePeriodDays`, keeping the chart in step
    ///    with the Statistics summary and the Today card.
    ///
    /// 2. **Abstinence.** A bucket that still holds the open day is not a
    ///    *completed* period, so it must never be flagged abstinent. The green
    ///    tick promises "this whole period was recorded alcohol-free", and that
    ///    claim cannot be settled while today is still running: the day could yet
    ///    become a drink day before the day-change time.
    ///
    /// The PDF export passes `nil`, so historical reports count every calendar day
    /// and every zero bucket is a finished, abstinent one.
    ///
    /// - Parameters:
    ///   - summaries: Daily totals for days that have entries.
    ///   - from: Inclusive period start, `yyyy-MM-dd`.
    ///   - to: Inclusive period end, `yyyy-MM-dd`.
    ///   - granularity: Bucket width; see `granularityForSpan(days:)`.
    ///   - inProgressDay: Optional current logical day.
    /// - Returns: Buckets in chronological order, or empty when `from > to` or a
    ///   date fails to parse.
    public static func bucketize(
        summaries: [DaySummary],
        from: String,
        to: String,
        granularity: ChartGranularity,
        inProgressDay: String? = nil
    ) -> [ChartBucket] {
        guard let start = DayResolver.parseDate(from),
              let end = DayResolver.parseDate(to),
              start <= end
        else { return [] }

        // O(1) lookup of a day's total; days absent here are abstinent (0 g).
        var gramsByDate: [String: Double] = [:]
        for summary in summaries { gramsByDate[summary.date] = summary.totalGrams }

        // One day past the period end; the exclusive upper bound when summing.
        let endExclusive = addingDays(1, to: end)

        var buckets: [ChartBucket] = []
        var bucketStart = start

        while bucketStart <= end {
            // Natural (un-clamped) end of this bucket, exclusive.
            let naturalEndExclusive: Date
            switch granularity {
            case .daily:
                naturalEndExclusive = addingDays(1, to: bucketStart)
            case .weekly:
                naturalEndExclusive = addingDays(7, to: bucketStart)
            case .monthly:
                // Month buckets snap to the 1st of the next month, so successive
                // buckets align to calendar months even when the first one starts
                // mid-month.
                naturalEndExclusive = firstOfNextMonth(after: bucketStart)
            }

            // Never let a bucket run past the period.
            let cappedEndExclusive = min(naturalEndExclusive, endExclusive)

            var sum = 0.0
            var dayCount = 0
            var day = bucketStart
            while day < cappedEndExclusive {
                sum += gramsByDate[DayResolver.formatDate(day)] ?? 0.0
                dayCount += 1
                day = addingDays(1, to: day)
            }

            // See the two consequences documented above.
            var bucketHoldsInProgressDay = false
            if let inProgressDay, let inProgress = DayResolver.parseDate(inProgressDay) {
                let isInBucket = inProgress >= bucketStart && inProgress < cappedEndExclusive
                let isDrinkDay = (gramsByDate[inProgressDay] ?? 0.0) > 0.0
                bucketHoldsInProgressDay = isInBucket
                if isInBucket && !isDrinkDay && dayCount > 0 { dayCount -= 1 }
            }

            buckets.append(
                ChartBucket(
                    labelDate: DayResolver.formatDate(bucketStart),
                    avgPerDay: dayCount > 0 ? sum / Double(dayCount) : 0.0,
                    // Abstinent = recorded fully alcohol-free AND a completed
                    // period. The guard is consequence (2): the current
                    // day/week/month never earns a tick until it closes.
                    isAbstinent: sum == 0.0 && !bucketHoldsInProgressDay
                )
            )

            // Advancing to the capped end (not the natural one) is what makes the
            // loop terminate on the final, possibly short, bucket.
            bucketStart = cappedEndExclusive
        }

        // Invariant: the loop always advances, so a non-empty range yields at
        // least one bucket.
        assert(!buckets.isEmpty, "bucketize: non-empty range produced no buckets")
        return buckets
    }

    // ── Calendar helpers ─────────────────────────────────────────────────────

    private static func addingDays(_ days: Int, to date: Date) -> Date {
        utcCalendar.date(byAdding: .day, value: days, to: date) ?? date
    }

    /// The first day of the month following `date`'s month, at the same noon
    /// anchor. Equivalent to Kotlin's `withDayOfMonth(1).plusMonths(1)`.
    private static func firstOfNextMonth(after date: Date) -> Date {
        var parts = utcCalendar.dateComponents([.year, .month], from: date)
        parts.day = 1
        parts.hour = 12
        guard let firstOfThisMonth = utcCalendar.date(from: parts),
              let next = utcCalendar.date(byAdding: .month, value: 1, to: firstOfThisMonth)
        else { return addingDays(1, to: date) }
        return next
    }
}
